Replace RCTLocalAssetImageLoader to RCTBundleAssetImageLoader

hellohublot opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory. -->

## Summary:

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

From the video below, we can see that the UI thread has dropped many frames, and it would become worse if there are multiple images.

If an image is located in the sandbox of the disk, we cannot load it using `RCTLocalAssetImageLoader` because `RCTLocalAssetImageLoader.requiresScheduling` is set to true, which loads the data on the UI thread and causes main thread stuttering. This will affect libraries such as `react-native-code-push` and others that save images to the sandbox of the disk.

Therefore, we should replace `RCTLocalAssetImageLoader.canLoadImageURL` from `RCTIsLocalAssetURL(url)` to `RCTIsBundleAssetURL(url)`. Similarly, we should rename the entire `RCTLocalAssetImageLoader` file with `RCTBundleAssetImageLoader`, which ignores images in the disk sandbox. And finally these images will be loaded from `NSURLRequest`, and our UI thread will run smoothly again.


https://user-images.githubusercontent.com/20135674/236368418-8933a2c6-549c-40d3-a551-81b492fe41d5.mp4




## Changelog:

<!-- Help reviewers and the release process by writing your own changelog entry.

Pick one each for the category and type tags:

[ANDROID|GENERAL|IOS|INTERNAL] [BREAKING|ADDED|CHANGED|DEPRECATED|REMOVED|FIXED|SECURITY] - Message

For more details, see:
https://reactnative.dev/contributing/changelogs-in-pull-requests
-->

[IOS] [Breaking] - Replace `RCTLocalAssetImageLoader` to `RCTBundleAssetImageLoader`

## Test Plan:

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->


Test Code:
```javascript
constructor(props) {
  super(props)
  this.state = {
    bundle_image: require('./large_image.png'),
    sandbox_image: '',
    source: null,
    isLoading: false,
  }
}

render() {
  console.log('render', this.state)
  return (
    <View style={{ flex: 1, padding: 50, backgroundColor: 'white'}}>
      <View style={{ flexDirection: 'row', alignItems: 'center', height: 70}}>
      {
        [{ title: 'Save Image To SandBox', onPress: () => {
          let image = Image.resolveAssetSource(this.state.bundle_image)
          console.log(image.uri)
          this.setState({ isLoading: true })
          RNFetchBlob.config({ fileCache: true, appendExt: "png" }).fetch("GET", image.uri).then(response => {
            let path = response.path()
            path = /^file:\/\//.test(path) ? path : 'file://' + path
            console.log(path)
            this.state.sandbox_image = path
          }).finally(() => this.setState({ isLoading: false }))
        }}, { title: 'Load From SandBox', onPress: () => {
          this.setState({ source: { uri: this.state.sandbox_image } })
        }}, { title: 'Clear', onPress: () => {
          this.setState({ 
              source: null,
              isLoading: false
          })
        }}, { title: 'Load From Bundle', onPress: () => {
          this.setState({ source: this.state.bundle_image })
        }}].map((item, index) => {
          return (
              <Pressable
                  key={index}
                  style={{ height: '100%', justifyContent: 'center', flex: 1, borderWidth: 1, borderColor: 'black', marginLeft: index > 0 ? 15 : 0 }}
                  onPress={item.onPress}
              >
                  <Text style={{ textAlign: 'center' }}>{item.title}</Text>
              </Pressable>
          )
        })
      }
      </View>
      <ActivityIndicator style={{ marginTop: 10 }} animating={this.state.isLoading} />
      <Image 
        key={`${this.state.source}`}
        style={{ marginTop: 20, width: 200, height: 200 }} 
        source={this.state.source}
        onProgress={({ nativeEvent }) => console.log(nativeEvent)}
        onLoadStart={() => this.setState({ isLoading: true })}
        onLoadEnd={() => this.setState({ isLoading: false })}
      />
    </View>
  )
}

```
It needs to be tested in three environments: [Simulator_Debug, RealDevice_Debug, RealDevice_Release]

1. Open `Perf Monitor` (RealDevice_Release can be skipped)
2. Click `Save Image to SandBox`
3. Wait for the loading to end and click `Load From SandBox`
4. Verify that the image can be loaded successfully
5. Verify that the `UI thread` keeps `60 FPS` (RealDevice_Release can be skipped)
6. Click `Clear`
7. Repeat steps [3, 4, 5, 6] several times
8. Click `Load From Bundle` to verify that the bundle image can be loaded successfully


Simulator_Debug

https://user-images.githubusercontent.com/20135674/236369344-ee1b8ff1-2d49-49f3-a322-d973f4adf3e7.mp4

RealDevice_Debug

https://user-images.githubusercontent.com/20135674/236369356-fe440b2b-f72a-49be-b63c-b4bf709dac8c.mp4

RealDevice_Release

https://user-images.githubusercontent.com/20135674/236369365-8a6a5c2f-09ad-4c90-b6bd-41e8a5e3aa7f.mp4

hellohublot commented:
@dmytrorykun  Hello, could you give me some suggestions for my PR? This issue has indeed persisted for a long time

cipolleschi commented on packages/react-native/Libraries/AppDelegate/RCTAppSetupUtils.mm:
```diff
@@ -16,7 +16,7 @@
 #import <React/RCTImageLoader.h>
 #import <React/RCTJSIExecutorRuntimeInstaller.h>
-#import <React/RCTLocalAssetImageLoader.h>
+#import <React/RCTBundleAssetImageLoader.h>
```
This is a huge breaking change and something that will break the internal system badly as it is.

My suggestion here is to:
1. create a copy of `RCTLocalAssetImageLoader` and rename it `RCTBundleAssetImageLoader`
2. Deprecate the `RCTLocalAssetImageLoader` for one version, at least, so that other libraries that may use this class won't break in the next release and will have some time to migrate away from it
3. Replace the usage in the `RCTAppSetupUtils` as you already did.

In this way, we can bring in the change more easily as it is more conservative. The change will require two passes but it will be much less disruptive.

cipolleschi commented on packages/react-native/Libraries/Image/RCTImagePlugins.h:
```diff
@@ -33,7 +33,7 @@ Class RCTGIFImageDecoderCls(void) __attribute__((used));
 Class RCTImageLoaderCls(void) __attribute__((used));
 Class RCTImageStoreManagerCls(void) __attribute__((used));
-Class RCTLocalAssetImageLoaderCls(void) __attribute__((used));
+Class RCTBundleAssetImageLoaderCls(void) __attribute__((used));
```
This file is automatically [generated by an internal script](https://github.com/facebook/react-native/pull/37232/files#diff-2f1b719a9f0e2a007a0a0d2eb6cfe34db6c40025b48419a3324c054525d26b78R7), so please, revert this change. 😅

cipolleschi commented on packages/react-native/Libraries/Image/RCTImagePlugins.mm:
```diff
@@ -23,7 +23,7 @@ Class RCTImageClassProvider(const char *name) {
     {"ImageLoader", RCTImageLoaderCls},
     {"ImageStoreManager", RCTImageStoreManagerCls},
-    {"LocalAssetImageLoader", RCTLocalAssetImageLoaderCls},
+    {"BundleAssetImageLoader", RCTBundleAssetImageLoaderCls},
```
Same as above: https://github.com/facebook/react-native/pull/37232/files#diff-02f0a739170ba7b5bf4bf6d64ef87c1fe840a13e486b289d78b74f4256006fc1L7

hellohublot commented:
@cipolleschi  Thanks, I have modified it according to your suggestion. At the same time, I also modified `RCTLocalAssetImageLoader.canLoadImageURL`, because although we replaced `RCTAppSetupUtils`, when `RCTImageLoader._loadersProvider` is empty, `RCTImageLoader._loaders` will still import `RCTLocalAssetImageLoader`, so for accurate testing, I also modified `RCTLocalAssetImageLoader.canLoadImageURL`, since it will be deprecated, so I guess we can modify it

hellohublot commented:
This CI error `test_ios_rntester_hermes_xcode_integration` is because of `RCTLocalizedString.h`, the `FBcoreLocalexxHash48` and `RCTLocalizedStringFromKey`

cipolleschi reviewed (approved):
Thank you for spending some time, making sure that this won't break anything.

cipolleschi commented:
/rebase

hellohublot commented:
/rebase

hellohublot commented:
@cipolleschi  Thanks. All the checks have been passed.

cipolleschi commented:
@dmytrorykun could you reimport the PR, please?

hellohublot commented:
@dmytrorykun Hello. Bump^

hellohublot commented:
@cipolleschi It has been reimported, Is there anything else I can help here? please feel free to ask
